examples/pytorch/gcn/train.py

```diff
@@ -1,145 +1,97 @@
-import argparse
-import time
-import numpy as np
 import torch
+import torch.nn as nn
 import torch.nn.functional as F
 import dgl
+import dgl.nn as dglnn
 from dgl.data import CoraGraphDataset, CiteseerGraphDataset, PubmedGraphDataset
+import argparse
 
-from gcn import GCN
-#from gcn_mp import GCN
-#from gcn_spmv import GCN
-
-
-def evaluate(model, features, labels, mask):
+class GCN(nn.Module):
+    def __init__(self, in_size, hid_size, out_size):
+        super().__init__()
+        self.layers = nn.ModuleList()
+        # two-layer GCN
+        self.layers.append(dglnn.GraphConv(in_size, hid_size, activation=F.relu))
+        self.layers.append(dglnn.GraphConv(hid_size, out_size))
+        self.dropout = nn.Dropout(0.5)
+
+    def forward(self, g, features):
+        h = features
+        for i, layer in enumerate(self.layers):
+            if i != 0:
+                h = self.dropout(h)
+            h = layer(g, h)
+        return h
+    
+def evaluate(g, features, labels, mask, model):
     model.eval()
     with torch.no_grad():
-        logits = model(features)
+        logits = model(g, features)
         logits = logits[mask]
         labels = labels[mask]
         _, indices = torch.max(logits, dim=1)
         correct = torch.sum(indices == labels)
         return correct.item() * 1.0 / len(labels)
 
 
-def main(args):
+def train(g, features, labels, masks, model):
+    # define train/val samples, loss function and optimizer
+    train_mask = masks[0]
+    val_mask = masks[1]
+    loss_fcn = torch.nn.CrossEntropyLoss()
+    optimizer = torch.optim.Adam(model.parameters(), lr=1e-2, weight_decay=5e-4)
+
+    # training loop
+    for epoch in range(200):
+        model.train()
+        logits = model(g, features)
+        loss = loss_fcn(logits[train_mask], labels[train_mask])
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+        acc = evaluate(g, features, labels, val_mask, model)
+        print("Epoch {:05d} | Loss {:.4f} | Accuracy {:.4f} "
+              . format(epoch, loss.item(), acc))
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--dataset", type=str, default="cora",
+                        help="Dataset name ('cora', 'citeseer', 'pubmed').")
+    args = parser.parse_args()
+    print(f'Training with DGL intrinsic graph convolution module.')
+ 
```

Comment:
You might want to first create a transform object and then pass it to the dataset classes below. Also you need to compose `RemoveSelfLoop` and `AddSelfLoop` as some datasets have self-loops at the beginning for some of the nodes.

Reply:
Yes, I can do that (creating a transform object `transform=AddSelfLoop()` and pass trans. obj to dataset class)
I think, by default (`allow_duplicate=False`), `AddSelfLoop` includes `RemoveSelfLoop` to remove duplicated self-loops, if I understand the doc correctly (https://docs.dgl.ai/generated/dgl.transforms.AddSelfLoop.html).

Reply:
I see. Thanks.